### ISSUE TITLE
[FEA] Add libcudf functionality to support sampling from dataframes

### ISSUE THREAD

**Issue by RFinkelberg** [rapidsai/cudf#1879]
**Is your feature request related to a problem? Please describe.**
When implementing a cudf/cuml version of [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html), you need a way to shuffle a dataframe (or its indices). The idiomatic pandas way of doing this is [pandas.DataFrame.sample](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html). Generally we should support a native way of sampling randomly from dataframes.

**Describe the solution you'd like**
A method mimicking pandas functionality, such that
```python
df = cudf.DataFrame({'x': [1, 2, 3, 4]})

df.sample()
>>> 1

df.sample(3)
>>> [2, 4, 1]

df.sample(frac=0.5)
>>> [3, 4]
``` 

**Describe alternatives you've considered**
A functional way to do that is 
```python
df.iloc[numpy.random.permutation(df.index)]
```
but numba does not currently support this numpy function in a kernel, so it scales poorly.

**Comment by jrhemstad** [comment]
Sounds like this could be satisfied via a libcudf function to shuffle a column and/or table.

Something like:

```
table shuffle(table const& input)
```

Will create a new table by randomly shuffling all the rows of the input table.

Would that satisfy the need?

**Comment by RFinkelberg** [comment]
Functionally yes, since randomly sampling 3 elements is the same as shuffling and taking the first 3. I worry about the performance though. Will there be too much overhead shuffling the entire table? Would it be better to instead shuffle the indices and let the indexer deal with it?

**Comment by jrhemstad** [comment]
> Functionally yes, since randomly sampling 3 elements is the same as shuffling and taking the first 3. I worry about the performance though. Will there be too much overhead shuffling the entire table? Would it be better to instead shuffle the indices and let the indexer deal with it?

Yeah, I assumed you'd want to just shuffle the index for your use case, but proposed a more generic API that could be used for other things as well.

**Comment by RFinkelberg** [comment]
> That said, that would still require shuffling the entire index. This would be less generic, but more efficient:

I actually think that's okay, and closer to what we want. Shuffling a length `n` table reduces to sampling it `n` times without replacement.

**Comment by cjnolet** [comment]
I've been thinking over sampling without replacement primitive for cuML. We have need this functionality in cuML for several different algorithms- shuffling for mini-batch SGD & pasting for Random Forests are two that immediately come to mind. 

This could be accomplished using thrust by generating a uniformly random vector of size `n` and sorting it by key, where the indices of the original vector are sorted by the randomly generated values. If some number of samples are needed that are smaller than `n`, let's call it `m`, you can just take the first `m` values from the resulting sorted array.

**Comment by jrhemstad** [comment]
Closing as a duplicate of https://github.com/rapidsai/cudf/issues/1426
